[spec] remove implementation details from enum spec (#18579)
lampepfl/dotty: docs/_spec/05-classes-and-objects.md

- We use `<...>` for syntactic constructs that in some circumstances might be empty.
For instance, `<value-params>` represents one or more parameter lists `(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)` or nothing at all.
- Enum classes fall into two categories:
  - _parameterized_ enum classes have at least one or more (possibly empty) term parameter clauses, denoted as `(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)`.

>   - _unparameterized_ enum classes have no term parameter clauses, but may optionally have a type parameter clause, denoted as `[´\mathit{tps}\,´]`.

Reviewer: That does not appear to be true:
```
scala> enum Foo { case Bar[T] }
-- [E004] Syntax Error: --------------------------------------------------------
1 |enum Foo { case Bar[T] }
  |                ^^^
  |                A case class must have at least one parameter list
  |
  | longer explanation available when compiling with `-explain`
1 error found
```

Author: this paragraph is about the base enum class, not a class case - hmm these terms are confusing

### Lowering of Enum Definitions

###### Summary
An enum class is represented as a `sealed abstract` class that extends the `scala.reflect.Enum` trait.

Enum cases are represented as follows:

> - a class enum case is mapped to a `case class` member of enum class' companion object,

Reviewer: ```suggestion
- a class enum case is mapped to a `case class` member of the enum class' companion object,
```

Author: done

> - a singleton enum case is mapped to a `val` member of the enum class' companion object, implemented by a local class definition which may be shared between cases.

Reviewer: ```suggestion
- a singleton enum case is mapped to a `val` member of the enum class' companion object, implemented by a local class definition. Whether that local class is shared with other singleton cases, and which ones, is left as an implementation detail.
```

Author: done

```
There are nine desugaring rules.
Rule (1) desugars enum definitions.
Rule (2) desugars cases of comma-separated names to simple enum cases.
Rules (3) to (7) desugar inferrable details of enum cases.
Rules (8) and (9) define how fully-desugared enum cases map into `case class`es or `val`s.
Explicit `extends` clauses must be provided in the following cases, where rules (2) to (6) do not apply:
- any enum case of a parameterized enum,
- any singleton enum case of an unparameterized enum with non-variant type parameters,
- any class enum case of an enum with type parameters, where the case also has type parameters.

1.  An `enum` definition
    ```scala

>     enum ´E´ <type-params> <value-params> { <defs> <cases> }

Reviewer: ```suggestion
    enum ´E´ <type-params> <value-params> extends <parents> { <defs> <cases> }
```

Author: done

    ```
    expands to a `sealed abstract` class that extends the `scala.reflect.Enum` trait and an associated companion object that contains the defined cases, expanded according to rules (2 - 8).
    The enum class starts with a compiler-generated import that imports the names `<caseIds>` of all cases so that they can be used without prefix in the class.
    ```scala
    sealed abstract class ´E´ <type-params> <value-params>
        extends <parents> with scala.reflect.Enum {
      import ´E´.{ <caseIds> }
      <defs>
    }
    object ´E´ { <cases> }
    ```

2. A simple enum case consisting of a comma-separated list of names
   ```scala
   case ´C_1´, ..., ´C_n´
   ```
   expands to the following simple enum cases
   ```scala
   case ´C_1´; ...; case ´C_n´
   ```
   Any modifiers or annotations on the original case extend to all expanded cases.

>    <p>This result is then further rewritten by either (3 or 4).</p>

Reviewer: Does that actually render in a meaningful way?

Author: I thought it looks nicer with the clean break, there's no way to do that in pure markdown without breaking the numbered list

4. A simple enum case  `´C´` of an unparameterized enum `´E´[´\mathit{tps}´]` with type parameters
   ```scala
   case ´C´
   ```
   where `´\mathit{tps}´` are of the following form
   ```scala
   ´\mathit{v}_1´ ´T_1´ >: ´L_1´ <: ´U_1´ ,   ... ,   ´\mathit{v}_n´ ´T_n´ >: ´L_n´ <: ´U_n´      (n > 0)
   ```
   and where each of the variances `´\mathit{v}_i´` is either `'+'` or `'-'`, expands to the following value enum case:
   ```scala
   case ´C´ extends ´E´[´B_1´, ..., ´B_n´]
   ```
   where `´B_i´` is `´L_i´` if `´\mathit{v}_i´ = '+'` and `´U_i´` if `´\mathit{v}_i´ = '-'`.
   <p>This result is then further rewritten with rule (8).</p>

5. A class enum case with type parameters, but without an extends clause
   ```scala
   case ´C´[´\mathit{tps}´](´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)
   ```
   of an unparameterized enum `´E´` without type parameters expands to
   ```scala
   case ´C´[´\mathit{tps}´](´\mathit{ps}_1\,´)...(´\mathit{ps}_n´) extends ´E´
   ```
   This result is then further rewritten with rule (9).

6. A class enum case without type parameters or an extends clause
   ```scala
   case ´C´(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)
   ```

>    of an unparameterized enum `´E´[´\mathit{tps}´]` with type parameters expands to

Author: we can work on the exact terminology, but these desugaring rules where the extends clause is inferred requires that the base enum has no term parameters - so I was calling this an "unparameterized enum". Maybe you could suggest a better term than "enum without term parameters, but has type parameters"

>    of an unparameterized enum `´E´[´\mathit{tps}´]` with type parameters expands to

Reviewer: Ah, OK, that's the same misunderstanding as above. It would be good to find a better name because "parameterized" refers to *type* parameters everywhere else in the spec. But it's not an issue with this PR's changes, so we can defer that.

>    of an unparameterized enum `´E´[´\mathit{tps}´]` with type parameters expands to

Author: ok then this is to revisit
